tests/app/dao/notification_dao/test_notification_dao_delete_notifications.py

```diff
@@ -357,29 +254,78 @@ def test_insert_notification_history_delete_notifications(sample_email_template)
+    create_notification(template=sample_email_template,
+                        created_at=datetime.utcnow() + timedelta(hours=1), status='delivered')
 
     # should NOT be deleted
     create_notification(template=sample_email_template,
-                        created_at=datetime.utcnow() - timedelta(minutes=59), status='temporary-failure')
+                        created_at=datetime.utcnow() + timedelta(minutes=61), status='temporary-failure')
     create_notification(template=sample_email_template,
                         created_at=datetime.utcnow() + timedelta(hours=1, seconds=1), status='temporary-failure')
-    create_notification(template=sample_email_template,
-                        created_at=datetime.utcnow() - timedelta(hours=1), status='delivered')
     create_notification(template=sample_email_template,
                         created_at=datetime.utcnow() + timedelta(minutes=20), status='created')
+    # should NOT be deleted - wrong status
     create_notification(template=sample_email_template,
                         created_at=datetime.utcnow() - timedelta(days=1), status='sending')
     create_notification(template=sample_email_template,
                         created_at=datetime.utcnow() - timedelta(days=1), status='technical-failure')
+    create_notification(template=sample_email_template,
+                        created_at=datetime.utcnow() - timedelta(hours=1), status='created')
 
     del_count = insert_notification_history_delete_notifications(
         notification_type=sample_email_template.template_type,
         service_id=sample_email_template.service_id,
-        start_time=datetime.utcnow(),
-        end_time=datetime.utcnow() + timedelta(hours=1))
+        timestamp_to_delete_backwards_from=datetime.utcnow() + timedelta(hours=1))
 
-    assert del_count == 3
+    assert del_count == 5
     notifications = Notification.query.all()
     history_rows = NotificationHistory.query.all()
-    assert len(history_rows) == 3
+    assert len(history_rows) == 5
     assert len(notifications) == 6
+
+
+def test_insert_notification_history_delete_notifications_more_notifications_than_query_limit(sample_template):
+    create_notification(template=sample_template,
+                        created_at=datetime.utcnow() + timedelta(minutes=4), status='delivered')
+    create_notification(template=sample_template,
+                        created_at=datetime.utcnow() + timedelta(minutes=20), status='permanent-failure')
+    create_notification(template=sample_template,
+                        created_at=datetime.utcnow() + timedelta(minutes=30), status='temporary-failure')
+
+    del_count = insert_notification_history_delete_notifications(
+        notification_type=sample_template.template_type,
+        service_id=sample_template.service_id,
+        timestamp_to_delete_backwards_from=datetime.utcnow() + timedelta(hours=1),
+        qry_limit=1
+    )
+
+    assert del_count == 1
```

Comment:
I'm confused with this test. Am I right in thinking that are 3 notifications that should be moved but they would only do so one by one as qry_limit is one. `insert_notification_history_delete_notifications` will only delete maximum `qry_limit` meaning that if we want to move all our history, we need to call `insert_notification_history_delete_notifications` multiple times until it's got through them all? I don't see this happening anywhere?

Reply:
This method should adhere to the query limit... but there should be another test for `delete_notifications_older_than_retention_by_type` to make sure if notification count is greater than query limit it will delete them all, this `test_delete_notifications_calls_subquery_multiple_times` should test that. This test is poorly named.

Reply:
Strange, I can see the test `test_delete_notifications_calls_subquery_multiple_times` which looks like it's testing the right thing, however I can't find the corresponding logic in the app which actually makes it happen. Could you point me at it please?

Reply:
Ah... there is a small problem here. This test works because it iterates over the day hour by hour. But we need another loop to complete the work before moving onto the next hour. 
I've now added it: `app/dao/notifications_dao.py:315`